kcauto.sikuli/combat.sikuli/lbas.py

```diff
@@ -146,28 +146,40 @@ def _check_and_manage_lbas_fatigue(self, fatigue, group):
                 group_fatigue['medium']
                 if group_fatigue['medium'] else fatigue['medium'])
         else:
+            Util.log_msg("LBAS group #{} has good morale.".format(group))
             lbas_group_nodes_key = 'lbas_group_{}_nodes'.format(group)
             # put LBAS group into air defense mode if it is active but no nodes
             # are assigned to it
             group_sortie_mode = (
                 'sortie'
                 if self.config.combat[lbas_group_nodes_key] else 'defense')
+            Util.log_msg("Assign LBAS group #{a} to {b} mode.".format(a=group, b=group_sortie_mode))
             self._switch_lbas_mode(group_sortie_mode)
         return fatigue
 
-    def _switch_lbas_mode(self, final_mode):
+    def _switch_lbas_mode(self, lbas_group_mode_name_target):
         """Switches the lbas group mode to the specified final mode.
 
         Args:
-            final_mode (str): the mode to switch the LBAS group to
+            lbas_group_mode_name_target (str): the mode to switch the LBAS group to
         """
         Util.rejigger_mouse(self.regions, 'top')
+        # Must be relevant to http://kancolle.wikia.com/wiki/Land_Base_Aerial_Support#Options
+        if sum([lbas_group_mode_name_target == i for i in self._lbas_group_modes_tuple]) != 1:
+            raise ValueError("No such LBAS mode: \"{}\".".format(lbas_group_mode_name_target))
         while not self.module_regions['lbas_mode_switcher'].exists(
-                'lbas_group_mode_{}.png'.format(final_mode)):
-            Util.click_preset_region(self.regions, 'lbas_mode_switch_button')
-            Util.rejigger_mouse(self.regions, 'top')
-            Util.kc_sleep(1)
-        Util.log_msg("LBAS group switched to {} mode.".format(final_mode))
+                'lbas_group_mode_{}.png'.format(lbas_group_mode_name_target), 1 + Globals.SLEEP_MODIFIER):
+            for lbas_group_mode_name_current in self._lbas_group_modes_tuple:
+                lbas_group_mode_pattern = Pattern('lbas_group_mode_{}.png'.format(
+                    lbas_group_mode_name_current)).similar(Globals.DEFAULT_SIMILARITY)
+                Util.rejigger_mouse(self.regions, 'top')
+                if self.module_regions['lbas_mode_switcher'].exists(
+                        lbas_group_mode_pattern, 1 + Globals.SLEEP_MODIFIER):
+                    print("Current LBAS group mode is \"{}\".".format(lbas_group_mode_name_current))
+                    if lbas_group_mode_name_current == lbas_group_mode_name_target:
```

Comment:
I like where you're going trying to remove the blind-click behavior from LBAS group assignment, but as it stands this is a really slow way to do it. This algorithm loops through every single possible matches after every click, until it actually finds the mode it's trying to find. With the limited size of the set (5) one might think this should be a fast operation, but each search is actually in the realm of significant milliseconds, not nanoseconds.

I think the appropriate approach would be this: reorder the values of the tuple so that they're actually in the order they show in-game. In this method, do a `for` search for which mode the squad is currently in. Then, create a re-ordered list that contains which modes we should expect to see on every click with something like this:
```python
expected_modes = self._lbas_group_modes_tuple[n:] + self._lbas_group_modes_tuple[:n]
```
Where `n` is the index of the matched current state. This will ensure that each subsequent click incurs an O(1) cost.

Reply:
Have you tested it already? I'll order the tuple, however, it's not slow to implement `CircularBuffer` class since the target `Region` is just tiny. Moreover, there is also fun related with `threading` or `multiprocessing` related implementations. And after this we will put here  `Util.kc_sleep()` call to beware of catbombs from fast clicking. Not a first priority, sorry.

Reply:
I did test this earlier and was getting a noticeable delay on 'standby' since it was the last one in the list, but this was on a pretty slow machine. `Thread`s are an option but I it's a big overkill. My proposal was aiming to avoid a major refactor.

Reply:
I didn't fix any of the linting errors but this is my take on the logic:

```python
    def _switch_lbas_mode(self, lbas_group_mode_name_target):
        """Switches the lbas group mode to the specified final mode.

        Args:
            lbas_group_mode_name_target (str): the mode to switch the LBAS group to
        """
        Util.rejigger_mouse(self.regions, 'top')
        # Must be relevant to http://kancolle.wikia.com/wiki/Land_Base_Aerial_Support#Options
        if sum([lbas_group_mode_name_target == i for i in self._lbas_group_modes_tuple]) != 1:
            raise ValueError("No such LBAS mode: \"{}\".".format(lbas_group_mode_name_target))
        for idx, lbas_group_mode_name_current in enumerate(self._lbas_group_modes_tuple):
            if self.module_regions['lbas_mode_switcher'].exists(
                    'lbas_group_mode_{}.png'.format(lbas_group_mode_name_current), 1 + Globals.SLEEP_MODIFIER):
                break
        expected_modes = self._lbas_group_modes_tuple[idx:] + self._lbas_group_modes_tuple[:idx]
        for idx, mode in enumerate(expected_modes):
            print("Current LBAS group mode is \"{}\".".format(lbas_group_mode_name_current))
            if mode == lbas_group_mode_name_target:
                break
            Util.check_and_click(self.module_regions['lbas_mode_switcher'], 'lbas_group_mode_{}.png'.format(mode))
            Util.rejigger_mouse(self.regions, 'top')
            self.module_regions['lbas_mode_switcher'].wait('lbas_group_mode_{}.png'.format(expected_modes[idx + 1]))
        Util.log_msg("LBAS group switched to {} mode.".format(lbas_group_mode_name_target))
```